src/Analyser/NodeScopeResolver.php

```diff
@@ -2597,6 +2597,13 @@ static function (): void {
 				$throwPoints[] = ThrowPoint::createImplicit($scope, $expr);
 			}
 
+			if (
+				$parametersAcceptor instanceof ClosureType && count($parametersAcceptor->getImpurePoints()) > 0
+				&& $scope->isInClass()
```

Comment:
last open question: do we want to invalidate 'this' in this case for regular function calls with side-effects?
we only handle Closure at this point right now

Reply:
Scratch that. A function is pretty unlikely to affect `$this` or `self`